src/components/install-app.tsx

```diff
@@ -0,0 +1,159 @@
+import { Workbox } from "workbox-window/index";
+import queryString from "query-string";
+import { getOfflineManifest, saveOfflineManifestToOfflineActivities,
+  cacheUrlsWithProgress } from "../offline-manifest-api";
+import { OfflineManifest } from "../types";
+import { OfflineManifestLoadingModal } from "./offline-manifest-loading-modal";
+import { queryValue } from "../utilities/url-query";
+
+interface IState {
+  serviceWorkerVersionInfo?: string;
+  loadingOfflineManifest: boolean;
+  offlineManifest?: OfflineManifest;
+}
+interface IProps {
+}
+
+export class InstallApp extends React.PureComponent<IProps, IState> {
+  private wb: Workbox;
+
+
+  public constructor(props: IProps) {
+    super(props);
+
+    if (!isOfflineHost() || !("serviceWorker" in navigator)) {
+      // Don't setup workbox
+      this.state = {
+        loadingOfflineManifest: false
+      };
+      // TODO: show some useful message incase some one loads this page from a
+      // non offline host, or in a browser without service worker support
+      return;
+    }
+
+    const offlineManifestId = queryValue("offlineManifest");
+    const loadingOfflineManifest = !!offlineManifestId;
+
+    this.state = {
+      serviceWorkerVersionInfo: "Starting...",
+      loadingOfflineManifest
+    };
+
+    this.wb = new Workbox("service-worker.js");
+
+    this.wb.addEventListener("waiting", (event) => {
+      // TODO: in future work we should show a dialog using this recipe:
+      // https://developers.google.com/web/tools/workbox/guides/advanced-recipes#offer_a_page_reload_for_users
+      // For now just send a message to skip waiting so we don't have to do it manually in the devtools
+      // This will trigger a 'controlling' event which we are listening for below and reload the page when
+      // we get it
+      this.wb.messageSkipWaiting();
+    });
+
+    navigator.serviceWorker.ready.then(registration => {
+      // Cache the application as soon as there is an active service worker
+      // it doesn't matter if the serivce worker is controlling the page because
+      // the fetches are done by the service worker not the page
+      this.cacheApplication();
+    });
+
+    this.wb.register().then((_registration) => {
+      console.log("Workbox register() promise resolved", _registration);
+
+      console.log("Sending GET_VERSION_INFO to service worker...");
+      this.setState({serviceWorkerVersionInfo: "Checking..."});
+      this.wb.messageSW({type: "GET_VERSION_INFO"})
+        .then((versionInfo: unknown) => {
+          if (typeof versionInfo === "string") {
+            this.setState({serviceWorkerVersionInfo: versionInfo});
+          } else {
+            this.setState({serviceWorkerVersionInfo: "Invalid response!"});
+          }
+        })
+        .catch(() => {
+          this.setState({serviceWorkerVersionInfo: "No response!"});
+        });
+    });
+
+    if (offlineManifestId) {
+      getOfflineManifest(offlineManifestId)
```

Comment:
Should this be done in `cacheApplication` so that we know the service worker is ready to cache it or if not can you add a comment about why it is called here instead of there?

Reply:
I moved it to cacheApplication and added more comments around it.